### PR TITLE
feat(native restaking): add a function to update the states of native token restaking

### PR DESCRIPTION
**Pull request by TimmyExogenous**
## Description

This PR adds a function to update the state when Oracle updates the balance of the native restaking token.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new method for iterating over undelegation records by staker and asset.
	- Added functionality to manage native restaking balances for stakers and assets.

- **Improvements**
	- Enhanced error handling across various functions to provide clearer messages and control flow.
	- Updated delegation operations to allow for more nuanced success and error reporting.
	- Renamed functions for improved clarity and readability.
	- Streamlined test preparation processes for better parameterization.

- **Bug Fixes**
	- Modified logic in share management to correctly reflect changes in operator shares.

- **Documentation**
	- Improved clarity of error messages for better debugging and user experience.

### CODE REVIEW

File: x/delegation/keeper/delegation_state.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-type DelegationOpFunc func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error
+type DelegationOpFunc func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) (bool, error)
 
 // IterateDelegationsForStakerAndAsset processes all operations
 // that require iterating over delegations for a specified staker and asset.
```
Comment (leonz789) [inline x/delegation/keeper/delegation_state.go]: Can we refactor `IterateDelegationsForStakerAndAsset` and `IterateDelegationsForStaker` into one method ?

File: x/delegation/keeper/delegation_state.go [hunk 2]
```diff
@@ -60,22 +66,22 @@ func (k Keeper) IterateDelegationsForStaker(ctx sdk.Context, stakerID string, op
...
 		if err != nil {
-			return err
+			return true, err
 		}
 		amount = amount.Add(singleAmount)
-		return nil
+		return false, nil
 	}
 	err = k.IterateDelegationsForStakerAndAsset(ctx, stakerID, assetID, opFunc)
 	if err != nil {
```
Comment (leonz789) [inline x/delegation/keeper/delegation_state.go]: these 4 lines equals to 
`return amount, err`

File: x/delegation/keeper/delegation_state.go [hunk 3]
```diff
@@ -60,22 +66,22 @@ func (k Keeper) IterateDelegationsForStaker(ctx sdk.Context, stakerID string, op
 // It needs to be calculated from the share and amount of the asset pool.
 func (k Keeper) StakerDelegatedTotalAmount(ctx sdk.Context, stakerID string, assetID string) (amount sdkmath.Int, err error) {
 	amount = sdkmath.NewInt(0)
-	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error {
+	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) (bool, error) {
 		if amounts.UndelegatableShare.IsZero() {
-			return nil
+			return true, nil
```
Comment (leonz789) [inline x/delegation/keeper/delegation_state.go]: `return false, nil`
iteration should move on to next operator

File: x/delegation/keeper/delegation_state.go [hunk 4]
```diff
@@ -88,19 +94,19 @@ func (k Keeper) StakerDelegatedTotalAmount(ctx sdk.Context, stakerID string, ass
 // the key of return value is the operator address, and the value is the asset amount.
 func (k *Keeper) AllDelegatedAmountForStakerAsset(ctx sdk.Context, stakerID string, assetID string) (map[string]sdkmath.Int, error) {
```
Comment (leonz789) [inline x/delegation/keeper/delegation_state.go]: The name is kind of confused with `StakerDelegatedTotalAmount`.
Maybe changed to some name like `DelegatedAmountPerOperatorForStakerAsset`.

File: x/delegation/keeper/delegation_state.go [hunk 5]
```diff
@@ -177,9 +183,9 @@ func (k *Keeper) GetSingleDelegationInfo(ctx sdk.Context, stakerID, assetID, ope
 func (k *Keeper) GetDelegationInfo(ctx sdk.Context, stakerID, assetID string) (*delegationtype.QueryDelegationInfoResponse, error) {
 	var ret delegationtype.QueryDelegationInfoResponse
 	ret.DelegationInfos = make(map[string]*delegationtype.DelegationAmounts, 0)
```
Comment (leonz789) [inline x/delegation/keeper/delegation_state.go]: equals to `make(string]*delegationtype.DelegationAmounts)`

File: x/delegation/keeper/un_delegation_state.go [hunk 6]
```diff
@@ -150,6 +150,37 @@ func (k *Keeper) GetStakerUndelegationRecords(ctx sdk.Context, stakerID, assetID
 	return k.GetUndelegationRecords(ctx, recordKeys)
 }
 
+// IterateUndelegationsByStakerAndAsset iterate the undelegation records according to the stakerID and assetID.
+func (k *Keeper) IterateUndelegationsByStakerAndAsset(
+	ctx sdk.Context, stakerID, assetID string, isUpdate bool,
+	opFunc func(undelegation *types.UndelegationRecord) (bool, error),
+) error {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixStakerUndelegationInfo)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(strings.Join([]string{stakerID, assetID}, "/")))
```
Comment (leonz789) [inline x/delegation/keeper/un_delegation_state.go]: use `delegationtype.GetDelegationStateIteratorPrefix(stakerID, assetID)` ?

File: x/delegation/keeper/update_native_restaking_balance.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+			return err
+		}
+	} else if amount.IsNegative() {
+		// If the balance decreases due to the client chain PoS slashing, the decreased amount
+		// will be slashed from the withdrawable amount first, the pending undelegation second,
+		// and the delegated share last if there is still a remaining amount that needs to be slashed.
+		// When slash from the delegated share, we can slash sequentially from all delegated operators
+		// according to the order of the operator addresses in the KV store. We can also proportionally
+		// slash from all operators to whom the related staker has already delegated, if we think
+		// this approach is better
```
Comment (MaxMustermann2) [inline x/delegation/keeper/update_native_restaking_balance.go]: I think this approach is better.

Say that native staked asset deposited by staker S is 64 ETH of which 32 ETH is free, 16 ETH is delegated to operator A and 16 ETH is unbonding from operator B. In this scenario, staker S is running a validator on Ethereum but may not necessarily be running one on Exocore or any of its AVSs.

The native staked ETH is slashed in half to 32 ETH. This half should be applied to each of the items:
- Free deposit -> 32 ETH becomes 16 ETH
- Delegated to A -> 16 ETH becomes 8 ETH and the vote power change is applied immediately across the AVSs in which A has opted.
- Unbonding from B -> 16 ETH becomes 8 ETH and continues unbonding according to the same schedule

File: x/delegation/keeper/delegation_state.go (hunk 1 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/delegation_state.go]: Thanks, I will keep these two interfaces and use a common function to refactor them.

File: x/delegation/keeper/delegation_state.go (hunk 2 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/delegation_state.go]: Thanks, I will refine it in the next commit.

File: x/delegation/keeper/delegation_state.go (hunk 3 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/delegation_state.go]: Thanks, I will fix it in the next commit.

File: x/delegation/keeper/delegation_state.go (hunk 4 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/delegation_state.go]: Thanks, I will rename them to `TotalDelegatedAmountForStakerAsset` and `AllDelegatedInfoForStakerAsset` in the next commit.

File: x/delegation/keeper/delegation_state.go (hunk 5 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/delegation_state.go]: right, I will change it in the next commit.

File: x/delegation/keeper/un_delegation_state.go (hunk 6 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/un_delegation_state.go]: Right, I will rename the function `GetDelegationStateIteratorPrefix` and use it here.

File: x/delegation/keeper/update_native_restaking_balance.go (hunk 7 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance.go]: I will adjust the implementation when slashing from the delegated amount.

File: x/delegation/keeper/update_native_restaking_balance_test.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	// check the asset states of two operators
+	defaultOperatorAsset, err := suite.App.AssetsKeeper.GetOperatorSpecifiedAssetInfo(suite.Ctx, suite.opAccAddr, assetID)
+	suite.NoError(err)
+	suite.Equal(sdkmath.NewInt(7), defaultOperatorAsset.TotalAmount)
+	suite.Equal(expectedShareForDefaultOpeartor, defaultOperatorAsset.TotalShare)
+
+	anotherOperatorAsset, err := suite.App.AssetsKeeper.GetOperatorSpecifiedAssetInfo(suite.Ctx, anotherOperatorAddr, assetID)
+	suite.NoError(err)
+	suite.Equal(sdkmath.NewInt(14), anotherOperatorAsset.TotalAmount)
+	suite.Equal(expectedShareForAnotherOpeartor, anotherOperatorAsset.TotalShare)
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance_test.go]: Seems like it's unnecessary to refactor.

File: x/delegation/keeper/update_native_restaking_balance_test.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	// check the undelegation state after slashing
+	records, err := suite.App.DelegationKeeper.GetStakerUndelegationRecords(suite.Ctx, stakerID, assetID)
+	suite.NoError(err)
+	suite.Equal(sdkmath.NewInt(0), records[0].ActualCompletedAmount)
+
+	// check the delegated share for two operators
+	delegationForDefaultOperator, err := suite.App.DelegationKeeper.GetSingleDelegationInfo(suite.Ctx, stakerID, assetID, suite.opAccAddr.String())
+	suite.NoError(err)
+	slashProportion := sdkmath.LegacyNewDec(1).Sub((sdkmath.LegacyNewDec(1)).Quo(sdkmath.LegacyNewDec(3)))
+	expectedShareForDefaultOpeartor := sdkmath.LegacyNewDec(10).Mul(slashProportion)
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance_test.go]: thanks

File: x/delegation/keeper/update_native_restaking_balance_test.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	suite.Equal(expectedShareForDefaultOpeartor, delegationForDefaultOperator.UndelegatableShare)
+
+	delegationForAnotherOperator, err := suite.App.DelegationKeeper.GetSingleDelegationInfo(suite.Ctx, stakerID, assetID, anotherOperatorAddr.String())
+	suite.NoError(err)
+	expectedShareForAnotherOpeartor := sdkmath.LegacyNewDec(20).Mul(slashProportion)
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance_test.go]: thanks

File: x/delegation/keeper/update_native_restaking_balance.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		if remainAmount.IsPositive() {
+			// calculate the slash proportion
+			totalDelegatedAmount, err := k.TotalDelegatedAmountForStakerAsset(ctx, stakerID, assetID)
+			if err != nil {
+				return err
+			}
+			slashProportion := sdkmath.LegacyNewDecFromBigInt(remainAmount.BigInt()).Quo(sdkmath.LegacyNewDecFromBigInt(totalDelegatedAmount.BigInt()))
+			if slashProportion.GT(sdkmath.LegacyNewDec(1)) {
+				slashProportion = sdkmath.LegacyNewDec(1)
+			}
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance.go]: Thanks, I will address this issue in the next commit.

File: x/operator/keeper/usd_value.go [hunk 12]
```diff
@@ -469,30 +469,30 @@ func (k *Keeper) CalculateUSDValueForStaker(ctx sdk.Context, stakerID, avsAddr s
...
 	}
 	totalUSDValue := sdkmath.LegacyNewDec(0)
-	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error {
+	opFunc := func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) (bool, error) {
 		if keys.OperatorAddr == operator.String() {
 			if _, ok := assets[keys.AssetID]; ok {
 				price, ok := prices[keys.AssetID]
 				if !ok {
-					return errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
+					return true, errorsmod.Wrapf(operatortypes.ErrKeyNotExistInMap, "CalculateUSDValueForStaker assetID doesn't exist, assetID:%s", keys.AssetID)
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/operator/keeper/usd_value.go]: Thanks, I will update it in the next commit.

File: x/operator/keeper/usd_value.go [hunk 13]
```diff
@@ -469,30 +469,30 @@
...
 				if err != nil {
-					return err
+					return true, err
 				}
 				usdValue := CalculateUSDValue(amount, price.Value, assetInfo.AssetBasicInfo.Decimals, price.Decimal)
 				totalUSDValue = totalUSDValue.Add(usdValue)
 			}
 		}
-		return nil
+		return false, nil
```
[the remark this answers is not in the document]
Comment (TimmyExogenous) [reply x/operator/keeper/usd_value.go]: Thanks, I will add the code comment in the next commit.

**Review by adu-web3 — APPROVED** [review]
LGTM

**Review by leonz789 — APPROVED** [review]

File: x/delegation/keeper/update_native_restaking_balance.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+		// whom the related staker has already delegated
+		if remainAmount.IsPositive() {
+			// calculate the slash proportion
+			totalDelegatedAmount, err := k.TotalDelegatedAmountForStakerAsset(ctx, stakerID, assetID)
+			if err != nil {
+				return err
+			}
+			if !totalDelegatedAmount.IsZero() {
+				slashProportion := sdkmath.LegacyNewDecFromBigInt(remainAmount.BigInt()).Quo(sdkmath.LegacyNewDecFromBigInt(totalDelegatedAmount.BigInt()))
+				if slashProportion.GT(sdkmath.LegacyNewDec(1)) {
```
Comment (bwhour) [inline x/delegation/keeper/update_native_restaking_balance.go]: Replace this magic numbers with constants

> const (
>     MaxSlashProportion = 1
> )

File: x/delegation/keeper/update_native_restaking_balance.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+					slashProportion = sdkmath.LegacyNewDec(1)
+				}
+				opFunc := func(keys *delegationtypes.SingleDelegationInfoReq, delegationAmount *delegationtypes.DelegationAmounts) (bool, error) {
+					opAccAddr, err := sdk.AccAddressFromBech32(keys.OperatorAddr)
+					if err != nil {
+						return true, err
+					}
+					slashShare := delegationAmount.UndelegatableShare.Mul(slashProportion)
+					actualSlashAmount, err := k.RemoveShare(ctx, false, opAccAddr, stakerID, assetID, slashShare)
```
Comment (bwhour) [inline x/delegation/keeper/update_native_restaking_balance.go:108]: can we extract reusable logic into standalone functions ? like 

> func (k Keeper) slashFromWithdrawable(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from withdrawable amount
}

>func (k Keeper) slashFromUndelegations(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from the pending undelegated amount.
}

>func (k Keeper) slashFromDelegatedShare(ctx sdk.Context, stakerID, assetID string, amount sdkmath.Int) (sdkmath.Int, error) {
    // Implement the logic to deduct from the delegated share.
}

File: x/delegation/keeper/update_native_restaking_balance.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+					if err != nil {
+						return true, err
+					}
+					err = k.assetsKeeper.UpdateStakerAssetState(ctx, stakerID, assetID, types.DeltaStakerSingleAsset{
+						TotalDepositAmount: actualSlashAmount.Neg(),
+					})
+					if err != nil {
+						return true, err
+					}
+					remainAmount = remainAmount.Sub(actualSlashAmount)
```
Comment (bwhour) [inline x/delegation/keeper/update_native_restaking_balance.go]: it is better to use _pendingSlashAmount_ to replace remainAmount

**Review by bwhour — APPROVED** [review]
Add some comment. the overall looks good.

File: x/delegation/keeper/update_native_restaking_balance.go (hunk 14 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance.go]: ok

File: x/delegation/keeper/update_native_restaking_balance.go (hunk 16 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance.go]: fixed

File: x/delegation/keeper/update_native_restaking_balance.go (hunk 15 above)
Comment (TimmyExogenous) [reply x/delegation/keeper/update_native_restaking_balance.go:108]: I will record this issue and change it in the future.

**Pull request merged**